mimxrt: Add the class PWM to the  machine module.

robert-hh opened the pull request:
Frequency range 15Hz to > 1 MHz, with decreasing resolution of
the duty cycle. The basic API is supported as documented, except
that keyword parameters are accepted for both the instantiation and
the PWM.init() call.

Extensions:
- Support PWM for channel pairs.

Channel pairs are declared by supplying 2-element tuples for the
pins. The two channels of a pair must be the A/B channel of a
FLEXPWM module. These form than a complementary pair.

Additional supported keywprd parameters:

center=value Defines the center position of a pulse within the
    pulse cycle. The align keyword is actually shortcut for center.

sync=True|False: If set to True, the channels will be synchronized to
    a submodule 0 channel, which has to be set up before.

align=PWM.MIDDLE | PMW.BEGIN | PWM.END. It defines,
    whether synchronized channels are Center-Aligned or Edge-
    aligned. The channels must be either complementary a channel pair or
    a group of synchronized channels. It may as well be applied to a
    single channel, but withiout any benefit.

invert= 0..3. Controls ouput inversion of the channles. Bit
    0 controls the first channel, bit 1 the second.

start=True|False start=False causes the PWM output to defer
    starting. The pwm.start() method can later be used to start a
    group of channels at the same time.

deadtime=time_ns time of complementary channels for delaying the rising
    slope.

xor=0|1|2 xor causes the output of channel A and B to be xor-ed.
    If applied to a  X channel, it shows the value of A ^ B.
    If applied to an A or B channel, both channel show the xor-ed signal
    for xor=1. For xor=2, the xor-ed signal is split between channels A
    and B. See also the Reference Manual, chapter about double pulses.
    The behavior of xor=2 can also be achieved using the center method
    for locating a pulse within a clock period.

Additional methods:

pwm.start([submodules]) and pwm.stop([submodules]): start or stop
a channel (pair) or submodule group.

pwm.center([value]): Set the center position of the pulse in the range
    1..65535.

robert-hh committed:
mimxrt: Support PWM using the FLEXPWM and QTMR modules.

Frequency range 15Hz/18Hz to > 1 MHz, with decreasing resolution of
the duty cycle. The basic API is supported as documentated, except
that keyword parameters are accepted for both the instatiaton and
the PWM.init() call.

Extensions:
- Support PWM for channel pairs.

Channel pairs are declared by supplying 2-element tuples for the
pins. The two channels of a pair must be the A/B channel of a
FLEXPWM module. These form than a complementary pair.

Additional supported keywprd parameters:

center=value Defines the center position of a pulse within the
  pulse cycle. The align keyword is actually shortcut for center.

sync=True|False: If set to True, the channels will be synchronized to
  a submodule 0 channel, which has already to be enabled.

align=PWM.MIDDLE | PMW.BEGIN | PWM.END. It defines,
  whether synchronized channels are Center-Aligned or Edge-
  aligned. The channels must be either complementary a channel pair or
  a group of synchronized channels. It may as well be applied to a
  single channel, but withiout any benefit.

invert= 0..3. Controls ouput inversion of the pins. Bit
  0 controls the first pin, bit 1 the second.

start=True|False start=False causes the PWM output to not start
  immediately. The pwm.start() method can later be used to start a
  group of channels at the same time.

deadtime=time_ns time of complementary channels for delaying the rising
  slope.

xor=0|1|2 xor causes the output of channel A and B to be xored.
  If applied to a    X channel, it shows the value oif A ^ B.
  If applied to an A or B channel, both channel show the xored signal
  for xor=1. For xor=2, the xored signal is split between channels A
  and B. See also the Reference Manual, chapter about double pulses.
  The behavior of xor=2 can also be achieved using the center method
  for locating a pulse within a clock period.

The output is enabled for board pins only.

CPU pins may still be used for FLEXPWM, e.g. as sync source, but the
signal will not be routed to the output. That applies only to FLEXPWM
pins. The use of QTMR pins which are not board pins will be rejected.

Recent activities:

- Implement the review comments.
- Fix PWM glitch for FlexPWM and QTMR
  The FLEXPWM glitch was caused by the function ticks_delay_us64(),
  which is called by sleep(), and sleep() was used in the test script.
  The cure consisted of removing a _WFE() statement, which caused
  already earlier trouble.

dpgeorge commented:
> Additional methods:
> 
> pwm.start([submodules]) and pwm.stop([submodules]): start or stop a channel (pair) or submodule group.
> 
> pwm.center([value]): Set the center position of the pulse in the range 1..65535.

Have these methods been removed since the initial posting of this PR?  I seem to remember doing a first-pass review of this... in general the API of this PWM class should match the generic API provided by `extmod/machine_pwm.c`, which it looks like it does.

robert-hh commented:
Yes. These methods have been removed (sigh).

dpgeorge commented on ports/mimxrt/boards/MIMXRT1010_EVK/mpconfigboard.mk:
```diff
@@ -8,6 +8,7 @@ MICROPY_HW_FLASH_SIZE ?= 0x1000000  # 16MB
 
 JLINK_PATH ?= /media/RT1010-EVK/
 JLINK_COMMANDER_SCRIPT = $(BUILD)/script.jlink
+BOARD_QTMR_SUPPORT = 0
```
Is this an existing config value or a new one?  If new, can it be something like `MICROPY_HW_QTMR_SUPPORT`, to match others?

robert-hh replied:
OK. I changed it. It has a kind of double negation logic, such that it only has to be defined for the MIMXRT1011 devices.

dpgeorge commented on ports/mimxrt/hal/pwm_backport.c:
```diff
@@ -0,0 +1,198 @@
+ * Copyright 2016-2017 NXP *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * The MIT License (MIT)
```
I'm not sure about making a file have 2 licenses... I think if you copied and adapted the code from the NXP library then it should remain just BSD-3-Clause, and you can add your copyright to that if you like (ie don't add the MIT text).

robert-hh replied:
OK. Yes, I took that code from NXP, because the NXP lib did not support duty_u16, and I did not want to loose resolution by scaling up from percent to u16.

dpgeorge commented on ports/mimxrt/hal/pwm_backport.h:
```diff
@@ -0,0 +1,70 @@
+ * Copyright 2016-2017 NXP *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * The MIT License (MIT)
```
same thing re licensing as for the .c file

robert-hh replied:
OK

dpgeorge commented on ports/mimxrt/mpconfigport.h:
```diff
@@ -129,6 +129,11 @@ uint32_t trng_random_u32(void);
 #define MICROPY_PY_MACHINE_PULSE            (1)
+#define MICROPY_PY_MACHINE_PWM              (1)
+#define MICROPY_PY_MACHINE_PWM_INIT         (1)
+#define MICROPY_PY_MACHINE_PWM_DUTY         (1)
```
This option was intended just for legacy compatibility, ie existing ports that had `PWM.duty()`.  For new ports this option shouldn't be used, because there's `duty_u16` and `duty_ns` which should be plenty.

(What exactly did you make `duty()` do?)

robert-hh replied:
duty() is obviously ambiguous. Here it is duty_percent with a 0..100 range. Other ports have a 0..1024 range,

dpgeorge commented:
> These methods have been removed (sigh).

Don't throw away the code though, we can add them eventually, it just needs to be done in a "controlled" manner so that all ports follow the same path.

robert-hh commented:
Thank you for the review. I implemented the suggested changes, rebased it to master and squashed all commits but the last one with the review changes.

dpgeorge commented on ports/mimxrt/Makefile:
```diff
@@ -209,6 +210,11 @@ ifeq ($(MICROPY_PY_MACHINE_SDCARD),1)
 SRC_HAL_IMX_C += $(MCU_DIR)/drivers/fsl_usdhc.c
 endif
 
+ifneq ($(MICROPY_HW_QTMR_SUPPORT),0)
```
It might be cleaner to do something like the following (which is what stm32 does in these cases):
```
ifeq ($(MCU_SERIES),$(filter $(MCU_SERIES),MIMXRT1015 MIMXRT1021 MIMXRT1052 MIMXRT1062))
...
endif
```

robert-hh replied:
Since there are not too many PR's open at the moment, I would prefer adding MICROPY_HW_QTMR_SUPPORT=1 to the config files of the respective boards. That avoids a hiccup if another MCU type is added. Unfortunately one cannot extract easily a setting from a header file into the Makefile. Otherwise the information is in the MCU-Specific feature header file.

dpgeorge replied:
Yes, sounds good (to add `MICROPY_HW_QTMR_SUPPORT=1` where applicable).

robert-hh replied:
I changed my mind into following your suggestion. There is always more than one way to forget things, and it does not make a difference where to change if another MCU is added.
b.t.w.: We have both another board and MCU in the queue, but it's not ready yet. The board is SOM Vision, MIMXRT1062 or MIMXRT1052 based, where a user confirmed that the Teensy 4.1 build works in principle. The MCU is MIMXRT1176 with a MIMXRT1170_EVK board. The port progressed quite nicely, but there are still hiccups. But including this MCU/board requires changing the nxp-driver and tinyusb lib, and that is the more difficult part.

dpgeorge replied:
> But including this MCU/board requires changing the nxp-driver and tinyusb lib, and that is the more difficult part

We can do that, as long as it is standard tinyusb, ie just a newer version.

robert-hh commented:
I just noticed that changing the frequency or duty cycle inserts a small glitch in the signal at mimxrt1052 and mimxrt1062, but not the MIMXRT1021. Maybe that is the reason I did not see than before. According to the data sheet, that should not happen, at least not for changing the duty cycle. 
So put that PR on hold. I want to analyze that first and see, if I can change that.

dpgeorge commented:
> So put that PR on hold. I want to analyze that first and see, if I can change that.

Ok, so this PR is on hold.

robert-hh commented:
I could fix the glitch for the FLEXPWM channels. The culprit was at a completely different place in the file ticks.c, function ticks_delay_us64(), a _WFE() statement, which cased a stall of the PWM for up to 50 µs. ticks_delay_us64() was though sleep() in the test script.
The glitch in QTMR channels is still present. I happens when changing the configuration, e.g. with the freq() and duty_xx() methods. Glitches are in acy case expected at large changes of the frequency, which require changing the PWM module's input clock. But that is common to PWM implementations in other port as well, so we could give it a go.

robert-hh commented:
The glitches on the QTMR channels are now gone as well. So it's well suited now for merge.

dpgeorge commented on ports/mimxrt/hal/pwm_backport.c:
```diff
@@ -0,0 +1,198 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
```
This text (lines 10-26) is from the MIT license and should not be added.  The `SPDX-License-Identifier: BSD-3-Clause` line is enough to specify the license.

I would suggest the following:
```
/*
 * This file is part of the MicroPython project, http://micropython.org/
 *
 * Copyright (c) 2015, Freescale Semiconductor, Inc.
 * Copyright 2016-2017 NXP *
 * Copyright (c) 2021 Robert Hammelrath
 * 
 * SPDX-License-Identifier: BSD-3-Clause
 *
 */
```

robert-hh replied:
Thanks. Changed.

dpgeorge commented on ports/mimxrt/hal/pwm_backport.h:
```diff
@@ -0,0 +1,69 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
```
Same comment as for the .c file above.

robert-hh replied:
Changed as well.

robert-hh committed:
mimxrt: Change copyright notice.

Also rebased on the current master.

robert-hh commented:
Hi Damien. The Copyright Notice is changed accordingly, and the PR is rebased.

dpgeorge commented:
Thanks for the work on this, a good feature to have.  Merged in 81f706aee46d60aa8a91c524da839f743e4cc608
